### ISSUE TITLE
C# Unicode Operators

### ISSUE THREAD

**Issue by mikaeldui** [dotnet/roslyn#41241]
**Version Used**: 
3.1

**Steps to Reproduce**:
1. Use Unicode operators instead of the normal, like ≥ instead of >=.
2. Try compile.
3. See the error.

**Expected Behavior**:
**Unicode operators should be supported** as Unicode is widely available and limited support has already been introduced in C#.

**Actual Behavior**:
C# doesn't recognize Unicode operators and only compiles if one's using ASCII alternatives.

**Comment by CyrusNajmabadi** [comment]
> Unicode operators should be supported as Unicode is widely available

Unicode operators are not widely supported.  There are practically no keyboards that offer them.

> This means more energy is spent reading

This is definitely not the case.  These operators are ubiquitous across widespread programming languages.  Being different here would require more energy to understand as it would go against the trillions of lines of code out there that are consistent here in not using unicode.

> and writing code

That's also not hte case as you'd have to go through much more complicated chords to generate these versus just typing the simple chars on hte keyboard.

Note: if you'd like to see something like ≥ in code you can do so with teh right font.  For example, this is how things look to me:

![image](https://user-images.githubusercontent.com/4564579/73220298-eaa32a80-4112-11ea-90b2-2e4e00e19649.png)

Note that this is simply as aspect of my font/editor, and does not affect anyone else out there.

**Comment by Grauenwolf** [comment]
How would this work from an IDE perspective? I don't know how to even enter that symbol outside of a math equation editor.

**Comment by miloush** [comment]
I think the fact that some people don't have the means or knowledge of typing something is not a very good argument against supporting it, but I agree that Unicode being widely available argument is not much valid either.

I think the questions are 1) if someone submitted a pull request to add such support in addition to the existing operators, would it likely be rejected, 2) are there other, more favourable ways of allowing such scenario, such as user defined operators.

Nevertheless, I too would be interested if the request is only on the rendering side, in which case a font as suggested will do, or what scenario would such support unblock.

**Comment by CyrusNajmabadi** [comment]
> I think the fact that some people don't have the means... of typing something is not a very good argument against supporting it, 

To me, it's a very good reason.  If 99.99% of keyboards out there will make this complex and difficult, it's not a good thing for the language.  The language has to be practical for the ecosystem.

**Comment by miloush** [comment]
I would understand that if the proposal was to switch to the these operators and no longer support the old ones, but I might need some help understanding why it is a concern even when you can keep typing the same way as before.

**Comment by tannergooding** [comment]
It makes the language more complex; may prohibit certain users or languages from easily interacting with a codebase (especially if enforced as a requirement), etc.

Fonts provide the visual niceness of unicode without breaking users who can't or won't be able to use them. It keeps things simpler overall to leave it as external to the language.

**Comment by Grauenwolf** [comment]
> I think the fact that some people don't have the means or knowledge of typing something is not a very good argument against supporting it

Not some, most. Actually I would say the "vast majority" of people would be unable to use this feature.

At this point I don't even have reason to believe that you have the ability to use this feature. How would you type the character?

And don't say an alt-sequence, because telling people to memorize Alt+0167 is not a viable strategy.

**Comment by CyrusNajmabadi** [comment]
> but I might need some help understanding why it is a concern even when you can keep typing the same way as before.

http://stronglyemergent.com/blog/2013/negative-100-points/

**Comment by Dreamescaper** [comment]
Just use font with ligatures, smth like that:
https://github.com/tonsky/FiraCode/wiki/Visual-Studio-Instructions

**Comment by miloush** [comment]
Don't get me wrong, I don't think @mikaeldui has provided a convincing use case in the issue description.

> Fonts provide the visual niceness
> Just use font with ligatures
@CyrusNajmabadi already suggested that. This might or might not address the issue @mikaeldui is having. If yes, great, if not, I think we need more details.

> -100 points
I am aware of that too. The model has changed a bit for C# since then though, as it doesn't have to be the team investing all the time and effort into writing specification, implementing and testing it. I definitely agree there are more important things for the team to invest in.

> may prohibit certain users or languages from easily interacting with a codebase
Sure, if you enforce such style. I believe that should be a call of the codebase owner though.

> How would you type the character?
On a keyboard. Next to left shift on [this one](https://github.com/DandelionSprout/norsk-utvidet), next to right shift on [this](http://www.kreativekorp.com/software/keyboards/superlatin/) or (this)[http://www.kreativekorp.com/software/keyboards/supersymbol/] one for example. Or you can create your own.

**Comment by CyrusNajmabadi** [comment]
> On a keyboard. Next to left shift on this one, next to right shift on this or this one for example. Or you can create your own.

Right.  But the language isn't going to change to target these keyboards.  It's nonsensical to add that complexity to target such a tiny part of the userbase.

**Comment by CyrusNajmabadi** [comment]
> The model has changed a bit for C# since then though, as it doesn't have to be the team investing all the time and effort into writing specification, implementing and testing it.

Yes, the team still needs all that investment.  Because it still has to go through the full and complete pass from the team to get in.

**Comment by Grauenwolf** [comment]
Teams.

It's not just the compiler team, you also need the IDE team, the other IDE team, the TFS team, the third party IDE team,  and who knows what else. Especially when it requires regression testing unicode up and down every dev stack.

**Comment by ufcpp** [comment]
I agree with https://github.com/dotnet/roslyn/issues/41241#issuecomment-579912858.

Since Japanese OS has IME (Input Method), we can easily input non-ASCII characters including math symbols. For instance, type ふとうごう (which means "not-equal sign" in Japanese) and use IME, then you can get ≠, ≤, ≥, ≦, ≧, ≨, ≴, and so on. However almost all people in Japan don't want to use non-ASCII characters in source codes except for comments and string literals because use of IME (turning on/off IME) is annoying.

**Comment by miloush** [comment]
> Teams

I disagree with this one. Unicode is already supported all the way through, you can not only have any Unicode characters in comments and strings, but also class and member names provided they are in the right categories. The only affected team therefore seems to be the compiler one, but I'm happy to be corrected.

**Comment by miloush** [comment]
>  the language isn't going to change to target these keyboards

There is also the emoji panel.

![image](https://user-images.githubusercontent.com/10546952/73431980-98901e00-4339-11ea-8b09-07130c167c6a.png)

But that's not the point, anyone can create their own input method.

I don't think anyone is asking for a language to target any specific keyboard, it shouldn't.  You cannot type `<` and `>` on at least 60 of the keyboards Windows is shipping with anyway, so clearly no size of keyboard userbase would make a difference.

**Comment by CyrusNajmabadi** [comment]
> The only affected team therefore seems to be the compiler one, but I'm happy to be corrected.

You are incorrect.  It would require teh compiler to be updated.  The IDE.  The docs.  All teh analyzers.  All the refactorings.  The debugger.  All the features.  Etc. etc. etc.

> There is also the emoji panel.

This is not relevant.  It's not on the keyboards used by 99.9% of developers.  It's not a good idea for the language to have built-in knowledge of such things.

--

> you can not only have any Unicode characters in comments and strings, but also class and member names

This isn't actually relevant here.  None of those have any meaning to the language itself.  i.e. it's part of an identifier, but not part of any sort of language construct that everyone would be expected to use.  i.e. we don't expect any part of the core language to use anything out of ascii.  All keywords are ascii (and we're not going to accept unicode variants of them).  Operators, punctuation, braces, etc. are all easily written with ascii in a universal manner that works across all tooling *across the entire world*.

**Comment by miloush** [comment]
> This is not relevant.

I was answering the question how these particular characters can be typed other than Alt+number combinations (which they even cannot). The panel can be invoked all keyboards.

> are all easily written with ascii [...] across the entire world.

Unless you are in the part of the world that doesn't have those on their keyboards, then you are asking them to use a different one. And I am fine with that.

> we don't expect any part of the core language to use anything out of ascii. All keywords are ascii (and we're not going to accept unicode variants of them).

I think that answer is enough to close the issue.

> it's part of an identifier, but not part of any sort of language construct

That's why I tried to suggest an alternative way through user operators - they would be treated as identifiers and could have infix syntax, as discussed e.g. in https://github.com/dotnet/csharplang/issues/594.

Looks like we both agree on the conclusion but not the arguments and I think there isn't much point spending time convincing each other. :)

**Comment by gafter** [comment]
The team that is most affected by a change such as this suggested one is the *user community*.  Every addition to the language makes the language more difficult to learn for nearly 100% of its potential user base.  If that effort across its user base a not paid back by a corresponding improvement in usability for that user base, then the change is a net negative for the ecosystem.  It does not matter who would implement the feature or how much effort that implementation takes.

I am skeptical this change would result in an improvement in usability.

**Comment by gafter** [comment]
Since this is a language feature request, moving to csharplang.
